### PR TITLE
Customizable Controller Config

### PR DESCRIPTION
**Pull request by PoisonPoisonPoison**
Allows you to change the controls Individually through a txt file.
Limited to 100 entrys because its just editing the 360 configuration in memory.
Popups will show the correct button prompt.

### CODE REVIEW

File: Default_controls.txt [hunk 1]
```diff
@@ -0,0 +1,100 @@
+MAPPING_SPRINT BUTTON_CROSS
```
Comment (TheOfficialFloW) [inline Default_controls.txt:1]: Please move this file to `gamefiles` and call it `default_controls.txt`.

File: controls.txt [hunk 2]
```diff
@@ -0,0 +1,100 @@
+MAPPING_SPRINT BUTTON_CROSS
```
Comment (TheOfficialFloW) [inline controls.txt:1]: Move this to `gamefiles` and call it `alternative_controls.txt`.

File: loader/config.c [hunk 3]
```diff
@@ -38,34 +38,206 @@ int read_config(const char *file) {
...
-    CONFIG_VAR(use_shader_cache);
-    CONFIG_VAR(skygfx_ps2_shading);
-    CONFIG_VAR(skygfx_colorfilter);
-    CONFIG_VAR(skygfx_ps2_sun);
-    CONFIG_VAR(disable_detail_textures);
-    CONFIG_VAR(disable_ped_spec);
-    CONFIG_VAR(disable_tex_bias);
-    CONFIG_VAR(disable_mipmaps);
-    CONFIG_VAR(aa_mode);
+    #define 	CONFIG_VAR(var) if (strcmp(name, #var) == 0) config.var = value;
```
Comment (TheOfficialFloW) [inline loader/config.c:47]: Please revert the diff here.

File: loader/config.c [hunk 4]
```diff
@@ -38,34 +38,206 @@ int read_config(const char *file) {
...
+
+typedef struct {
+  char *name;
+  uint8_t action;
+} NameToButtonID;
+
+static NameToButtonID name_to_button_id[] = {
+  { "BUTTON_CROSS", BUTTON_CROSS },
+  { "BUTTON_CIRCLE", BUTTON_CIRCLE },
+  { "BUTTON_SQAURE", BUTTON_SQAURE },
```
Comment (TheOfficialFloW) [inline loader/config.c:83]: BUTTON_SQUARE

File: loader/config.c [hunk 5]
```diff
@@ -38,34 +38,206 @@ int read_config(const char *file) {
...
+  { "BUTTON_TRINAGLE", BUTTON_TRINAGLE },
```
Comment (TheOfficialFloW) [inline loader/config.c:84]: BUTTON_TRIANGLE

File: loader/main.c [hunk 6]
```diff
@@ -605,11 +629,17 @@ void patch_game(void) {
   kuKernelCpuUnrestrictedMemcpy((void *)so_find_addr("_ZN6CCheat16m_aCheatHashKeysE"), CCheat__m_aCheatHashKeys, sizeof(CCheat__m_aCheatHashKeys));
   hook_thumb(so_find_addr("_ZN6CCheat8DoCheatsEv"), (uintptr_t)CCheat__DoCheats);
 
+  // load custom 360 config
+  if (config.custom_controls)
+  {
```
Comment (TheOfficialFloW) [inline loader/main.c:634]: Remove { and }.

File: loader/main.c [hunk 7]
```diff
@@ -497,6 +497,30 @@ int MainMenuScreen__OnExit(void) {
 extern void *__cxa_guard_acquire;
 extern void *__cxa_guard_release;
 
+
+void patchControl(uint8_t btn, uint8_t action)
```
Comment (TheOfficialFloW) [inline loader/main.c:501]: Use 2 spaces instead of tabs.

File: loader/config.h [hunk 8]
```diff
@@ -22,6 +23,142 @@ enum SkyGfxColorFilter {
   SKYGFX_COLOR_FILTER_PS2,
   SKYGFX_COLOR_FILTER_PC,
 };
+enum Mappings{
```
Comment (TheOfficialFloW) [inline loader/config.h:26]: Use 2 spaces instead of tabs.

**Comment by TheOfficialFloW** [comment]
Thanks. However, before I merge this, please fix the style.

File: loader/config.h [hunk 9]
```diff
@@ -22,6 +23,142 @@ enum SkyGfxColorFilter {
...
+	MAPPING_FLIGHT_PRIMARY_ATTACK,
+	MAPPING_FLIGHT_SECONDARY_ATTACK,
+	MAPPING_FLIGHT_ALT_LEFT,
+	MAPPING_FLIGHT_ALT_RIGHT,
+	MAPPING_FLIGHT_ALT_UP,
+	MAPPING_FLIGHT_ALT_DOWN,
+	MAPPING_BASKETBALL_SHOOT,
+	MAPPING_BUNNY_HOP,
+	MAPPING_MAP_ZOOM_IN,
+	MAPPING_MAP_ZOOM_OUTHID_,
```
Comment (TheOfficialFloW) [inline loader/config.h:124]: Shouldn't those be named `HID_MAPPING_.*`?

File: loader/main.c (hunk 7 above)
Comment (TheOfficialFloW) [reply loader/main.c:501]: Actually, I will probably rewrite the hook myself.

**Comment by TheOfficialFloW** [comment]
Out of curiosity, did you extract the mapping manually or algorithmically?

**Comment by TheOfficialFloW** [comment]
I'll merge and fix up the code. Thanks for the PR!

**Comment by PoisonPoisonPoison** [comment]
I'm sorry, saw the messages too late.
The Mappings were in the binary I just copied them out and used n++ to automatically format them.

File: loader/config.c (hunk 4 above)
Comment (PoisonPoisonPoison) [reply loader/config.c:83]: akwward

Comment (PoisonPoisonPoison) [reply loader/config.c:83]: awkward

File: loader/config.c (hunk 5 above)
Comment (PoisonPoisonPoison) [reply loader/config.c:84]: awkward^2

**Pull request merged**